npk/src/manifest.rs

```diff
@@ -94,38 +179,58 @@ impl Manifest {
     }
 
     fn verify(&self) -> Result<(), Error> {
-        // TODO: check for none on env, autostart, cgroups, seccomp
-        if self.init.is_none() && self.args.is_some() {
+        // Most optionals in the manifest are not valid for a resource container
+        if self.init.is_none()
+            && (self.args.is_some()
+                || self.env.is_some()
+                || self.autostart.is_some()
+                || self.cgroups.is_some()
+                || self.seccomp.is_some()
+                || self.capabilities.is_some()
+                || self.suppl_groups.is_some()
+                || self.io.is_some())
+        {
             return Err(Error::Invalid(
-                "Arguments not allowed in resource container".to_string(),
+                "Resource containers must not define any of the following manifest entries:\
+                args, env, autostart, cgroups, seccomp, capabilities, suppl_groups, io"
+                    .to_string(),
             ));
         }
 
-        // The autostart option is only valid for startable containers
-        if self.autostart.is_some() && self.init.is_none() {
+        // Check for invalid uid or gid of 0
+        if self.uid == 0 {
+            return Err(Error::Invalid("Invalid uid of 0".to_string()));
+        }
+        if self.gid == 0 {
+            return Err(Error::Invalid("Invalid gid of 0".to_string()));
+        }
+
+        // Check for relative bind mounts
+        let sorted_bind_mounts: Vec<&PathBuf> = self
+            .mounts
+            .iter()
+            .filter(|(_, m)| matches!(m, Mount::Bind(_)))
+            .map(|(p, _)| p)
+            .sorted()
+            .collect();
```

Comment:
We can safe the collect by directly iterating over the iterator returned from `sorted` and do the below relative check in there as well.

Reply:
We need `sorted_bind_mounts` for two checks: The relative mount check and the overlapping mount check. That's why it is saved.

Reply:
But both checks can be done with one iteration. This would safe the additional copy.